File: app/src/main/java/com/ayan/ritual/data/Habit.kt
package com.ayan.ritual.data

import java.time.LocalDate
import java.time.Year

/**
 * One practice, and every day it was kept.
 *
 * Days are stored as epoch-days: a single Long per completion, timezone-free
 * and trivially cheap to hold a decade of.
 */
data class Habit(
    val id: String,
    val name: String,
    /** When in the day this belongs — shown above the name on every card. */
    val slot: String,
    val accentIndex: Int,
    val createdEpochDay: Long,
    val done: Set<Long>,
    /**
     * Who this makes them, in their own words: "Ayan is someone who reads
     * regularly". Shown verbatim wherever the practice appears, because a
     * sentence someone wrote about themselves is the thing they came for and
     * the task underneath it is only the evidence.
     *
     * Stored as typed rather than as a predicate to be re-conjugated. Three
     * grammars for one fact reads well in a mockup and breaks on the first
     * person who writes a sentence nobody anticipated.
     */
    val identity: String = "",
    /**
     * Day one of the thirty. Usually the day it was created, but its own
     * field so a goal can be taken again after a lapse without rewriting
     * history.
     */
    val goalStartEpochDay: Long = createdEpochDay,
    /** The day the thirty were cleared, or null while it is still running. */
    val builtEpochDay: Long? = null
) {
    fun isDone(date: LocalDate): Boolean = done.contains(date.toEpochDay())

    /** Days completed in [year], as day-of-year numbers, ready for the grid. */
    fun daysOfYear(year: Int): Set<Int> {
        val first = LocalDate.of(year, 1, 1).toEpochDay()
        val last = LocalDate.of(year, 12, 31).toEpochDay()
        val out = HashSet<Int>()
        for (day in done) {
            if (day in first..last) out.add((day - first).toInt() + 1)
        }
        return out
    }

    fun totalIn(year: Int): Int = daysOfYear(year).size

    /** Consecutive days ending today — or yesterday, so a day in progress doesn't read as broken. */
    fun streak(today: LocalDate): Int {
        var day = today.toEpochDay()
        if (!done.contains(day)) day -= 1
        var n = 0
        while (done.contains(day)) {
            n++
            day--
        }
        return n
    }

    /** True when yesterday was missed and today is not yet marked — Mochi's cue. */
    fun missedYesterday(today: LocalDate): Boolean =
        !done.contains(today.minusDays(1).toEpochDay()) && !isDone(today)

    fun bestStreak(): Int {
        if (done.isEmpty()) return 0
        val sorted = done.sorted()
        var best = 1
        var run = 1
        for (i in 1 until sorted.size) {
            run = if (sorted[i] == sorted[i - 1] + 1) run + 1 else 1
            if (run > best) best = run
        }
        return best
    }

    /* ── The thirty days ────────────────────────────────────────────────
       Everything below reads the same window: thirty days from the start,
       of which twenty four must be kept. Six may be missed, and six is said
       out loud everywhere rather than eighty per cent, because a budget you
       are spending is something a person can hold in their head and a grade
       you are failing is not. */

    /** Days of the goal that have already happened, 0 before it starts. */
    fun goalElapsed(today: LocalDate): Int =
        ((today.toEpochDay() - goalStartEpochDay) + 1)
            .coerceIn(0L, Goal.DAYS.toLong()).toInt()

    /** Days kept inside the window, whatever was marked outside it. */
    fun goalKept(): Int {
        val last = goalStartEpochDay + Goal.DAYS - 1
        return done.count { it in goalStartEpochDay..last }
    }

    /** Days spent from the allowance. Never negative, never over the cap. */
    fun goalMissed(today: LocalDate): Int =
        (goalElapsed(today) - goalKept()).coerceAtLeast(0)

    /** What is left of the six. Zero means the next miss ends it. */
    fun allowanceLeft(today: LocalDate): Int =
        (Goal.ALLOWED_MISSES - goalMissed(today)).coerceAtLeast(0)

    /** True once the thirty are done and enough of them were kept. */
    fun goalMet(today: LocalDate): Boolean =
        goalElapsed(today) >= Goal.DAYS && goalKept() >= Goal.NEEDED

    /**
     * True when the allowance is spent and the goal can no longer be met.
     * Worth knowing early: there is no point counting down to a day that
     * cannot arrive. It costs the goal and never the grid.
     */
    fun goalOutOfReach(today: LocalDate): Boolean =
        goalMissed(today) > Goal.ALLOWED_MISSES

    /** Already claimed and on the shelf. */
    val isBuilt: Boolean get() = builtEpochDay != null

    /** Days kept in total, which is what the shelf and the story card count. */
    val totalDone: Int get() = done.size

    companion object {
        /** Squares still unwritten in [year] — the days after today. */
        fun remainingIn(year: Int, today: LocalDate): Int =
            if (today.year != year) {
                if (today.year < year) Year.of(year).length() else 0
            } else {
                Year.of(year).length() - today.dayOfYear
            }
    }
}

File: ios/Shared/Habit.swift
import Foundation

/**
 One practice, and every day it was kept.

 Days are stored as epoch-days: a single `Int` per completion, timezone-free
 and trivially cheap to hold a decade of.
 */
struct Habit: Identifiable, Equatable {

    let id: String
    var name: String
    /// When in the day this belongs — shown above the name on every card.
    var slot: String
    var accentIndex: Int
    var createdEpochDay: Int
    var done: Set<Int>

    func isDone(_ date: DayDate) -> Bool { done.contains(date.epochDay) }

    /// Days completed in `year`, as day-of-year numbers, ready for the grid.
    func daysOfYear(_ year: Int) -> Set<Int> {
        let first = DayDate(year: year, month: 1, day: 1).epochDay
        let last = DayDate(year: year, month: 12, day: 31).epochDay
        var out = Set<Int>()
        for day in done where day >= first && day <= last {
            out.insert(day - first + 1)
        }
        return out
    }

    func totalIn(_ year: Int) -> Int { daysOfYear(year).count }

    /// Every square ever filled, across every year kept.
    var totalAllTime: Int { done.count }

    /// The first year this ritual has anything in it — where the archive starts.
    var firstYear: Int {
        DayDate(epochDay: min(createdEpochDay, done.min() ?? createdEpochDay)).year
    }

    /// Consecutive days ending today — or yesterday, so a day in progress
    /// doesn't read as broken.
    func streak(_ today: DayDate) -> Int {
        var day = today.epochDay
        if !done.contains(day) { day -= 1 }
        var n = 0
        while done.contains(day) {
            n += 1
            day -= 1
        }
        return n
    }

    /// True when yesterday was missed and today is not yet marked — Mochi's cue.
    func missedYesterday(_ today: DayDate) -> Bool {
        !done.contains(today.epochDay - 1) && !isDone(today)
    }

    func bestStreak() -> Int {
        if done.isEmpty { return 0 }
        let sorted = done.sorted()
        var best = 1
        var run = 1
        for i in 1..<sorted.count {
            run = sorted[i] == sorted[i - 1] + 1 ? run + 1 : 1
            if run > best { best = run }
        }
        return best
    }

    /// Squares still unwritten in `year` — the days after today.
    static func remainingIn(_ year: Int, today: DayDate) -> Int {
        if today.year != year {
            return today.year < year ? DayDate.lengthOfYear(year) : 0
        }
        return DayDate.lengthOfYear(year) - today.dayOfYear
    }

    /// Shared shape for every card, so the widget can't drift from the app.
    func model(today: DayDate, year: Int) -> SlabModel {
        SlabModel(
            title: name,
            slot: slot,
            accent: accentAt(accentIndex),
            year: year,
            today: today,
            doneDaysOfYear: daysOfYear(year),
            streak: streak(today),
            totalDone: totalIn(year),
            remaining: Habit.remainingIn(year, today: today),
            doneToday: isDone(today),
            mood: Cat.moodFor(
                doneToday: isDone(today),
                streak: streak(today),
                missedYesterday: missedYesterday(today)
            )
        )
    }
}
